icu4c/source/common/uloc.cpp

```diff
@@ -1755,8 +1744,14 @@ uloc_getLanguage(const char*    localeID,
         localeID=uloc_getDefault();
     }
 
-    i=ulocimp_getLanguage(localeID, language, languageCapacity, NULL);
-    return u_terminateChars(language, languageCapacity, i, err);
+    CharString result = ulocimp_getLanguage(localeID, NULL, *err);
+    if (U_FAILURE(*err)) {
+        return 0;
+    }
+
+    int32_t reslen = result.length();
+    uprv_memcpy(language, result.data(), std::min(languageCapacity, reslen));
```

Comment:
ditto

We should really add a CharString method `extract(dest, capacity, errorCode)` like https://unicode-org.github.io/icu-docs/apidoc/released/icu4c/classicu_1_1UnicodeString.html#a6d9b3ff497c7f0abf84cea6946b1633b

I can make a PR after yours for that.

Reply:
> ditto

Done.

> We should really add a CharString method `extract(dest, capacity, errorCode)`

I hope that when all of this is done one day, such a method won't really be needed then.